ds0_updater.py

```diff
@@ -5,13 +5,36 @@
 import logger
 import conf
 
+# keep it super simple for now because we don't really implement --allowed-keys yet.
+def verify_latest_commit():
+    allowed_keys = ds0_utils.allowed_keys()
+
+    # ensure the latest commit is signed and verified
+    while True:
+        for key in allowed_keys:
+            logger.exe('gpg --import ' + key['file'], expectError=True)
+        output = logger.exe('git log --pretty="format:%G?" --show-signature HEAD^..HEAD')
+
+        if "Can't check signature" in output[0]:
+            logger.info('gpg keys cleared on startup. Trying again...')
+            continue
+
+        verified = False
+        for key in allowed_keys:
+            if key['rsa_check'] in output[0] and key['signature_check'] in output[0]:
+                verified = True
+        if not verified:
+            logger.error('Scripts using a non-signed commit. Please ensure commit is valid.')
+            logger.error('    If it was a missed signature, feel free to open a ticket at https://github.com/riptano/ComboAMI.')
+        break
+
 # Update the AMI codebase if it's its first boot
 if not conf.get_config("AMI", "CompletedFirstBoot"):
     # check if a specific commit was requested
     force_commit = ds0_utils.required_commit()
 
     # update the repo
-    logger.exe('git pull')
+    logger.exe('git pull ' + ds0_utils.repository())
```

Comment:
> It would be very nice if --repository could accept any refspec.

I agree. Since I don't think `npm` supports that though, we'll have to come up with syntax. How about "if it starts with `+` it's a refspec, otherwise it's an `npm`-compatible URL"?

As for implementation, I like your idea. Let's think (out loud) about implementation. We break down whatever we get to repository, branch/tag, and commit-id. We have defaults - `https://github.com/riptano/ComboAMI`, branch `2.5`, and none (=latest), respectively. We can then do something like:
```
git remote remove origin
git remote add origin <repository>
git pull
git checkout <branch/tag> or git checkout tags/<branch/tag> if unsuccessful
git reset --hard <commit-id>
```
I think that lets us maintain backwards compatibility for `--forcecommit` by relying on defaults.

Does that make sense? Am I missing something?

> Thanks a ton for doing this and for being so responsive to feedback.

Sure! :) My pleasure. You guys are responsible for fantastic software and I'm sure we're not the only one this AMI is saving a lot of time for.